components/webext-storage/src/sync/mod.rs

```diff
@@ -0,0 +1,165 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+mod incoming;
+mod outgoing;
+
+use serde_derive::*;
+use serde_json;
+use sync15::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use incoming::IncomingAction;
+
+type JsonMap = serde_json::Map<String, serde_json::Value>;
+
+// We use the same values as places. Note that some of these values are
+// duplicated in the schema.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[repr(u8)]
+pub enum SyncStatus {
+    Unknown = 0,
+    New = 1,
+    Normal = 2,
+}
+
+/// For use with `#[serde(skip_serializing_if = )]`
+#[inline]
+pub fn is_default<T: PartialEq + Default>(v: &T) -> bool {
+    *v == T::default()
+}
+
+// XXX - need to work out how to consolidate "payload" vs "bso", particularly
+// the timestamp, etc. For now, we represent this in ServerPayload, even though
+// that's never actually stored in this way.
+// XXX - this isn't going to work without other changes.
+#[derive(Debug, Serialize, Deserialize)]
+struct ServerPayload {
+    guid: SyncGuid,
+    ext_id: String,
+    #[serde(default, skip_serializing_if = "is_default")]
+    data: Option<String>,
+    #[serde(default, skip_serializing_if = "is_default")]
+    deleted: bool,
+    last_modified: ServerTimestamp,
+}
+
+// Perform a 2-way or 3-way merge, where the incoming value wins on confict.
+// XXX - this needs more thought, and probably needs significant changes.
+// Main problem is that it doesn't handle deletions - but to do that, we need
+// something other than a simple Option<JsonMap> - we need to differentiate
+// "doesn't exist" from "removed".
```

Comment:
Yes, we could do this, but TBH I'm not sure what the exact semantics are here - it's not documented, so we probably need to check how chrome behaves in these cases. I've just punted doing anything else here for now as hopefully @glasserc is tackling that as part of #2900

Reply:
Summarizing the discussion from that thread, it sounds like Chrome's behavior is too coarse, and not really what we want...but also, the Kinto "client wins" strategy isn't right for us, and it's more important to maintain compatibility with what we had before than with Chrome's implementation. Is that accurate?

Since we have the mirror copy, @rfk's suggestion makes sense!